CEGAR loop to abstract values

makaimann opened the pull request:
This PR adds initial support for a CEGAR loop that abstracts values. There are several limitations, but those can be removed with future PRs. In particular,

* it currently **only** supports `CegProphecyArrays<IC3IA>` as the backend prover, because that is the main motivation for this
* it currently abstracts **all** values. A future PR could allow a cutoff size

A few things I needed to change:
* since we are nesting CEGAR loops, we needed to specify which version of `cegar_abstract/cegar_refine` was running
* similarly, it was difficult to abstract `bad_` as-is because that happened before `initialize`. Instead, I just set `bad_` in the constructor initializer list

I ran into some unrelated issues here which are out-of-scope for this PR, but I left a TODO. The issue was with transferring the engine's view of an invariant to the original TS. We need to cache all the symbols, but it was not considering UFs. This came up because there are UFs in the invariant from the `CegProphecyArrays` abstraction.

ahmed-irfan commented on engines/cegar_values.cpp:
```diff
@@ -0,0 +1,309 @@
+      }
+    }
+  } else {
+    prover_ts_ = conc_ts_;
```
prover_ts_ is pointing to conc_ts_. Is it intentional?

ahmed-irfan replied:
It is not clear to me, what is happening here?

makaimann replied:
Ahh I'll leave a comment, it's a copy constructor so I'm just starting the `prover_ts_` with all the same behavior as `conc_ts_` before abstracting it.

makaimann replied:
It's an easy way to avoid manually copying over all the state variables etc... (although I had to do that in the functional case anyway)

ahmed-irfan replied:
I see. So, va will have the old ts reference. Right?

ahmed-irfan commented on engines/cegar_values.cpp:
```diff
@@ -0,0 +1,309 @@
+  UnorderedTermMap & abstracted_values_;
+};
+
+TransitionSystem create_fresh_ts(bool functional, const SmtSolver & solver)
```
how about putting this method in the ts.h as a static function?

makaimann replied:
Great idea! It didn't work nicely to put in `ts.h` because it doesn't know about `Relational` or `Functional` `TransitionSystem`, but I created a new utility file in 6c7b9bc066942633f5447df8c9c8ce10ee1ce7da.

ahmed-irfan commented on pono.cpp:
```diff
@@ -64,7 +61,9 @@ ProverResult check_prop(PonoOptions pono_options,
 
   std::shared_ptr<Prover> prover;
-  if (pono_options.ceg_prophecy_arrays_) {
+  if (pono_options.cegp_abs_vals_) {
```
we should check if the right options are set.
e.g. ceg-proph is set to true.

makaimann replied:
I check in the `make_cegar_values_prover` function. Do you think we should check here also?

ahmed-irfan replied:
then it is fine.

ahmed-irfan commented on engines/cegar_values.cpp:
```diff
@@ -0,0 +1,301 @@
+            from_cegval_solver_.transfer_term(equalities[i], BOOL);
+        // TODO this should be more modular
+        //      can't assume super::abs_ts_ is the right one to constrain
+        super::abs_ts_.add_constraint(transferred_eq);
```
this is true for ic3ia and ceg-proph, but not for other engines.

We should specialize the refinement of the subprover ts, similar to what is done in the ceg-proph code.

makaimann replied:
Good idea! Done in 182ddd9bb267fc0dac4027b638d0a593af15238a

makaimann commented:
Thanks for the review! I think I've addressed all the feedback so far. I'd also like to add a cutoff value for when to abstract the value, but since this is already a large PR I'll do that in a separate PR.

ahmed-irfan reviewed (approved):
LGTM.

Thanks for making the changes.

makaimann commented:
Thanks for the reviews!
